src/router_core/modules/edge_addr_tracking/edge_addr_tracking.c

```diff
@@ -0,0 +1,372 @@
+                    if(inlink->link->edge_context != 0) {
+                        qdr_addr_endpoint_state_t *endpoint_state = (qdr_addr_endpoint_state_t *)inlink->link->edge_context;
+                        if (qdrc_can_send_address(addr, endpoint_state->conn) ) {
+                            qdrc_endpoint_t *endpoint = endpoint_state->endpoint;
+                            qdrc_send_message(addr_tracking->core, addr, endpoint, true);
+                        }
+                    }
+                    inlink = DEQ_NEXT(inlink);
+                }
+            }
+            break;
+        }
+        case QDRC_EVENT_ADDR_BECAME_DEST : {
+            //
+            // This address transitioned from zero to one destination. If this address already had local destinations
+            //
+            qdr_link_ref_t *inlink = DEQ_HEAD(addr->inlinks);
+            //
+            // Every inlink that has an edge context must be informed of the appearence of this address.
+            //
+            while (inlink) {
+                if(inlink->link->edge_context != 0) {
+                    qdr_addr_endpoint_state_t *endpoint_state = (qdr_addr_endpoint_state_t *)inlink->link->edge_context;
+                    if (qdrc_can_send_address(addr, endpoint_state->conn) ) {
+                        qdrc_endpoint_t *endpoint = endpoint_state->endpoint;
+                        qdrc_send_message(addr_tracking->core, addr, endpoint, true);
+                    }
+                }
+                inlink = DEQ_NEXT(inlink);
+            }
+        }
+        break;
+
+        case QDRC_EVENT_ADDR_NO_LONGER_LOCAL_DEST : {
+            // The address no longer has any local destinations.
+            // If there are no remote destinations either, we have to tell the edge routers to delete their sender links
+            if (qd_bitmask_cardinality(addr->rnodes) == 0) {
+                qdr_link_ref_t *inlink = DEQ_HEAD(addr->inlinks);
+                //
+                // Every inlink that has an edge context must be informed of the disappearence of this address.
+                //
+                while (inlink) {
+                    if(inlink->link->edge_context != 0) {
+                        qdr_addr_endpoint_state_t *endpoint_state = (qdr_addr_endpoint_state_t *)inlink->link->edge_context;
+                        qdrc_endpoint_t *endpoint = endpoint_state->endpoint;
+                        qdrc_send_message(addr_tracking->core, addr, endpoint, false);
+                    }
+                    inlink = DEQ_NEXT(inlink);
+                }
+            }
+
+            break;
+        }
+        case QDRC_EVENT_ADDR_ONE_LOCAL_DEST: {
+            //
+            // This address transitioned from N destinations to one local dest
+            // If this address already has non-zero remote destinations, there is no need to tell the edge routers about it
+            //
+            if (qd_bitmask_cardinality(addr->rnodes) == 0) {
```

Comment:
This is an unnecessary check.  The cardinality of rnodes is, by definition, zero.